spec.html

```diff
@@ -11817,6 +11815,26 @@ <h1>GetGlobalObject ( ): an Object</h1>
         1. Return _currentRealm_.[[GlobalObject]].
       </emu-alg>
     </emu-clause>
+
+    <emu-clause id="sec-runcallercontext" type="abstract operation">
+      <h1>
+        RunCallerContext (
+          _value_: an ECMAScript language value or ~empty~,
+        ): a Completion Record
+      </h1>
+      <dl class="header">
+        <dt>description</dt>
+        <dd>It resumes the caller context (sending it _value_ as the resumption value), and waits for the result, if any.</dd>
+      </dl>
+      <emu-alg>
+        1. Let _genContext_ be the running execution context.
+        1. Remove _genContext_ from the execution context stack and restore the execution context that is at the top of the execution context stack as the running execution context.
+        1. Let _callerContext_ be the running execution context.
+        1. Resume _callerContext_ passing NormalCompletion(_value_). If _genContext_ is ever resumed again, let _result_ be the Completion Record with which it is resumed.
+        1. Assert: If control reaches here, then _genContext_ is the running execution context again.
```

Comment:
Couldn't we do something like this?

```suggestion
        1. Resume _callerContext_, passing NormalCompletion(_value_).
        1. Assert: If control reaches here, then _genContext_ is the running execution context again.
        1. Let _result_ be the Completion Record with which _genContext_ was resumed.
```

Reply:
Writing `Resume _callerContext_, passing NormalCompletion(_value_).` as a complete step suggests (to me) that you can do that `Resume` and then immediately proceed to the next step of the AO. (Note that the spec has `Resume`s that, as far as I can tell, work exactly that way, in ScriptEvaluation, ExecuteModule, and PerformEval.)

Instead, I think it's important to emphasize that *this* `Resume` causes execution of this AO to pause indefinitely (possibly forever), and the only way for this step to complete (and control to then proceed to the next step) is for `_genContext_` to be resumed again.

Reply:
I don't see how that's unclear from my suggestion. I still have the assert. I've just moved the alias declaration to its own step later.

Reply:
I think the spec should be clear about whether a given `Resume` is 'blocking' or not. If (as in your suggestion) we don't try to convey that in the `Resume` step itself, then we should do so in its immediate neighborhood. The Assert step *sort of* helps, but I think we can do better. So if moving the `Let` to its own step is important, then how about this:
```suggestion
        1. Resume _callerContext_ passing NormalCompletion(_value_).
        1. NOTE: The above step transfers control to _callerContext_ and pauses. The only way for it to un-pause and have control proceed to the subsequent steps in this algorithm is for _genContext_ to be resumed again, which might never happen.
        1. Assert: _genContext_ is the running execution context.
        1. Let _result_ be the Completion Record with which _genContext_ was just resumed.
```
I like that the Note is more explicit and explanatory.

Reply:
That works for me.

Reply:
Done.

Thinking more about the content of the Note, I still like that it's being conveyed to the reader, but I'm concerned that it's being conveyed in a Note, which you should be able to delete/ignore without normative effect.

So I think we need to:
- give a normative semantics for blocking and non-blocking `Resume`, and
- syntactically distinguish them somehow.

However, I'm okay with leaving all that to a later PR. (Note that it should be somewhat easier if PR #2664 is merged.)